com.unity.shadergraph/Editor/Generation/Processors/Generator.cs

```diff
@@ -97,56 +100,77 @@ void BuildShader()
                 }
             }
             else
             {
-                NodeUtils.DepthFirstCollectNodesFromNode(activeNodeList, m_OutputNode);
+                NodeUtils.DepthFirstCollectNodesFromNode(activeNodeList, m_OutputNode, IncludeSelf.Include, null, ignoreActiveState);
             }
 
-            var shaderProperties = new PropertyCollector();
+            var allShaderProperties = new PropertyCollector();
             var shaderKeywords = new KeywordCollector();
-            m_GraphData.CollectShaderProperties(shaderProperties, m_Mode);
+            m_GraphData.CollectShaderProperties(allShaderProperties, m_Mode);
             m_GraphData.CollectShaderKeywords(shaderKeywords, m_Mode);
 
             if (m_GraphData.GetKeywordPermutationCount() > ShaderGraphPreferences.variantLimit)
             {
                 m_GraphData.AddValidationError(m_OutputNode.objectId, ShaderKeyword.kVariantLimitWarning, Rendering.ShaderCompilerMessageSeverity.Error);
 
-                m_ConfiguredTextures = shaderProperties.GetConfiguredTexutres();
+                m_ConfiguredTextures = allShaderProperties.GetConfiguredTexutres();
                 m_Builder.AppendLines(ShaderGraphImporter.k_ErrorShader);
             }
 
+            // Collect properties from the nodes
             foreach (var activeNode in activeNodeList.OfType<AbstractMaterialNode>())
-                activeNode.CollectShaderProperties(shaderProperties, m_Mode);
+                activeNode.CollectShaderProperties(allShaderProperties, m_Mode);
 
-            // Collect excess shader properties from the TargetImplementation
+            // Collect properties from the TargetImplementation
             foreach (var target in m_Targets)
             {
                 // TODO: Setup is required to ensure all Targets are initialized
                 // TODO: Find a way to only require this once
                 TargetSetupContext context = new TargetSetupContext();
                 target.Setup(ref context);
 
-                target.CollectShaderProperties(shaderProperties, m_Mode);
+                // flag all properties as being required by this target
+                allShaderProperties.BeginTargetCollection(target);
+                target.CollectShaderProperties(allShaderProperties, m_Mode);
+                allShaderProperties.EndTargetCollection(target);
             }
 
+            // set the property collector to read only
+            // (to ensure no rogue target or pass starts adding more properties later..)
+            allShaderProperties.SetReadOnly();
+
             m_Builder.AppendLine(@"Shader ""{0}""", m_Name);
             using (m_Builder.BlockScope())
             {
-                GenerationUtils.GeneratePropertiesBlock(m_Builder, shaderProperties, shaderKeywords, m_Mode);
+                GenerationUtils.GeneratePropertiesBlock(m_Builder, allShaderProperties, shaderKeywords, m_Mode);
```

Comment:
allShaderProperties are used to generate the property block (no change here -- same as before)